### PR TITLE
Webserver - include css, js in index

### PR DESCRIPTION
**Pull request by Elkropac**
## Description:
I'm building device to control sun air heater. It will be in place without local wifi. I want it to provide webserver on build-in AP, to be able to control some relays and read temperatures.
This MR enables inclusion of css and js directly into index page, so javascript works without internet access.

My config looks like this (using file from esphome-docs):
```yaml
web_server:
  port: 80
  css_include: "../../../esphome-docs/_static/webserver-v1.min.css"
  js_include: "../../../esphome-docs/_static/webserver-v1.min.js"
```

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#459

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/web_server/__init__.py [hunk 1]
```diff
@@ -16,7 +17,9 @@
     cv.GenerateID(): cv.declare_id(WebServer),
     cv.Optional(CONF_PORT, default=80): cv.port,
     cv.Optional(CONF_CSS_URL, default="https://esphome.io/_static/webserver-v1.min.css"): cv.string,
+    cv.Optional(CONF_CSS_INCLUDE, default=""): cv.string,
```
Comment (OttoWinter) [inline esphome/components/web_server/__init__.py]:
```suggestion
    cv.Optional(CONF_CSS_INCLUDE, default=""): cv.file,
```

File: esphome/components/web_server/__init__.py [hunk 2]
```diff
@@ -16,7 +17,9 @@
...
     cv.Optional(CONF_JS_URL, default="https://esphome.io/_static/webserver-v1.min.js"): cv.string,
+    cv.Optional(CONF_JS_INCLUDE, default=""): cv.string,
```
Comment (OttoWinter) [inline esphome/components/web_server/__init__.py]:
```suggestion
    cv.Optional(CONF_JS_INCLUDE, default=""): cv.file,
```

File: esphome/components/web_server/__init__.py [hunk 3]
```diff
@@ -39,3 +42,18 @@ def to_code(config):
     if CONF_AUTH in config:
         cg.add(var.set_username(config[CONF_AUTH][CONF_USERNAME]))
         cg.add(var.set_password(config[CONF_AUTH][CONF_PASSWORD]))
+    if CONF_CSS_INCLUDE in config and config[CONF_CSS_INCLUDE] != "":
+        if os.access(config[CONF_CSS_INCLUDE], os.R_OK):
+            with open(config[CONF_CSS_INCLUDE], "r") as myfile:
+                cg.add(var.set_css_include(myfile.read()))
+        else:
+            raise EsphomeError("Option {}.{} is defined, but file {} is not readable."
```
Comment (OttoWinter) [inline esphome/components/web_server/__init__.py]: This should be done in the validation step - see the comments above for changing cv.string to cv.file.

File: esphome/components/web_server/__init__.py (hunk 1 above)
Comment (Elkropac) [reply esphome/components/web_server/__init__.py]: I have to use cv.file_ , is that ok?

**Comment by Elkropac** [comment]
Should i prepare some docs and tests for this options? I'm not sure, how to add tests, when this options needs some files not present in project

File: esphome/components/web_server/web_server.cpp [hunk 4]
```diff
@@ -175,10 +185,17 @@ void WebServer::handle_index_request(AsyncWebServerRequest *request) {
...
+                  "<h2>Debug Log</h2><pre id=\"log\"></pre>"));
+  if (this->js_include_ != nullptr) {
+    stream->print(F("<script type=\"text/javascript\">"));
+    stream->print(this->js_include_);
+    stream->print(F("</script>"));
+  } else {
+    stream->print(F("<script src=\""));
+    stream->print(this->js_url_);
+    stream->print(F("\"></script>"));
+  }
```
Comment (glmnet) [inline esphome/components/web_server/web_server.cpp]: Is it needed to only support either of these? cannot be both supported?

Comment (Elkropac) [reply esphome/components/web_server/web_server.cpp]: I guess, i can change "} else {" to standalone "if (strlen(this->js_url_) > 0)" 

So you can define empty js_url in config to skip internet access.

I will test it tomorrow. (also for css)

Comment (Elkropac) [reply esphome/components/web_server/web_server.cpp]: I guess, i can change "} else {" to standalone "if (strlen(this->js_url_) > 0)"

So you can define empty js_url in config to skip internet access.

I will test it tomorrow. (also for css)

**Comment by ilkarataev** [comment]
Hi please fix it issue with merge.I need this changes)

**Comment by Elkropac** [comment]
Sorry, I did not find any time for fun projects last month. 
I have pushed suggested change. Is it possible to merge it now? Thanks.

**Comment by glmnet** [comment]
Hey looks great, thanks for accepting my suggestion

**Comment by glmnet** [comment]
So I am testing this now, if I try this:
````
web_server:
  css_url: ""
  css_include: "../esphome-docs/_static/webserver-v1.min.css"
  js_url: ""
  js_include: "../esphome-docs/_static/webserver-v1.min.js"
```` 

it sometimes crashes the ESP, for some reason (the stack trace is not decoded, maybe due to Windows line endings something about unicode)

However if I comment both css lines or both js lines the thing works ok, it loads the page properly, I was suspecting of the watchdog so I tried `App.feed_wdt()` but that doesn't seem to fix it.

Another likely issue would be a buffer overrun, I really don't know how the ESP allocates the buffer to print all the css and js.

Did you test it with the same settings as me? I'm attaching the full config of what I'm testing for your reference.

````
esphome:
  name: test_local2
  platform: ESP8266
  board: nodemcuv2
  platformio_options:
    upload_speed: 921600

api:

ota:

logger:
  level: VERBOSE

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_pass
  output_power: 20.5dB
  manual_ip:
    static_ip: 10.0.1.199
    gateway: 10.0.1.1
    subnet: 255.0.0.0

sensor:
  - platform: dht
    model: AM2302
    pin: D5
    update_interval: 60s
    temperature:
      name: "Living Room Temperature"
    humidity:
      name: "Living Room Humidity"


web_server:
  css_url: ""
  css_include: "../esphome-docs/_static/webserver-v1.min.css"
  js_url: ""
  js_include: "../esphome-docs/_static/webserver-v1.min.js"

````

**Comment by Elkropac** [comment]
Hi, i'm running this on esp32 nodemcu. It was ok so far, but it was not powered on for long time. 
However I rememeber, it was rebooting, when I was testing it month ago and i think, it was because nothing was connecting to API, so it triggered reboot (no crash, just reboot). 

Can you please test with
```yaml
api:
  reboot_timeout: 0s
```

If everything goes well, I'll have to add this to example I added to documentation

**Comment by glmnet** [comment]
It's definitely not a timeout, I would see in the logs the reboot reason, also the api connection is working, I must have pointed out that it reboots when I request the web page from the browser only.

I just tested it with ESP32 and it seems to work ok.

So I believe we can 
1) add some code and docs to make this possible on ESP32 only or
2) try to get it working on ESP8266 too, 

Most likely a buffer overrun issue now.

**Comment by Elkropac** [comment]
Ok. I think, i have one lolin board with ESP8266 available for testing.

Is it possible to get around this problem by loading input file to array of shorter strings (how long?) and printing this array in loop?
Or is there some other way?

**Comment by glmnet** [comment]
I tried to figure it out too
The thing is HTTP protocol sends content-length at the beginning so I'm pretty sure the web server will put all in memory so it can measure it and then start streaming it.

I was looking at the docs here https://github.com/me-no-dev/ESPAsyncWebServer but did not find an easy solution right away. There is a "chunked" something but that will require more changes.

Another solution might be like "hosting" the files, adding request handlers for the css and js, instead of inlining them in the same request, that would require less structural changes IMO

**Comment by Elkropac** [comment]
Sooo, something like this? Can you please test it on ESP8266? I can't now. It works on ESP32

File: esphome/components/web_server/web_server.cpp [hunk 5]
```diff
@@ -175,13 +184,43 @@ void WebServer::handle_index_request(AsyncWebServerRequest *request) {
...
+
+  request->send(stream);
+}
+#endif
+
+#ifdef WEBSERVER_JS_INCLUDE
+void WebServer::handle_js_request(AsyncWebServerRequest *request) {
+  AsyncResponseStream *stream = request->beginResponseStream("text/javascript");
+  if (this->css_include_ != nullptr) {
```
Comment (glmnet) [inline esphome/components/web_server/web_server.cpp]:
```suggestion
  if (this->js_include_ != nullptr) {
```

**Comment by glmnet** [comment]
Looks good, I'll test on ESP8266 soon.
Any special reason for the `#define`? I believe we try not to use them much as they cause the whole project cpp files to be recompiled again when a define changes. Maybe @OttoWinter or @brandond knows better

**Comment by Elkropac** [comment]
> Looks good, I'll test on ESP8266 soon.
> Any special reason for the `#define`? I believe we try not to use them much as they cause the whole project cpp files to be recompiled again when a define changes. Maybe @OttoWinter or @brandond knows better

Ah, ok, sorry about that. I was inspired by code in web_server.cpp and i thought it's there to make actual used code (based on config) as small as possible. I will wait for @OttoWinter  response on that.

**Comment by glmnet** [comment]
I just tested it with esp8266 and seems to be working good. I’ll merge with docs when you update that it serves files instead of in lining  them. 
Thanks!

**Comment by Elkropac** [comment]
Ok, thanks for review. 

Is it ok to merge this branch with 14 commits or should I create new, with commits of final solution only?

**Comment by glmnet** [comment]
Ahh GitHub squashes all commits together so no need for anything else, thanks!!

**Comment by Elkropac** [comment]
Great, thanks.

Next, I planned to add support to change fan speed and monochromatic light  brightness (I use it as PWM) thru web interface for my project. 

If you are interested, I can create feature request for that, so we can discuss how it should look like.

**Comment by glmnet** [comment]
ahh yes, sure! I'll have to review the code how it works already (web server control) but certainly this will be useful.

**Comment by Elkropac** [comment]
I think, most of it is in place (api part, JS part). It's just about rendering some (meaningful) controls on webpage

**Pull request merged**